datalake: fix potential skipped leadership notification

andrwng opened the pull request:
Previously the following sequence seemed possible:
- a replica has partition translator of term 9
- leadership is lost for term 9
- stop_translator() is called and a task is scheduled to remove the translator from the map and stop it, but doesn't run right away
- the replica becomes leader for term 10
- the leadership notification runs for term 10 and does a lookup of the translators, but one is still there from term 9, so the notification is a no-op
- stop_translator() task completes, removing the translator from term 9
- term 10 has no translator and never gets another notification to start a new one

This commit tweaks the code to make the notifications synchronously add and remove translators from the map, rather than deferring that to background work. The actual stopping of the translators is still backgrounded, but that shouldn't be problematic.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

### Bug Fixes

* Fixes a race that could prevent Iceberg translation from happening following a leadership change.
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

andrwng committed:
datalake: fix potential skipped leadership notification

Previously the following sequence seemed possible:
- a replica has partition translator of term 9
- leadership is lost for term 9
- stop_translator() is called and a task is scheduled to remove the
  translator from the map and stop it, but doesn't run right away
- the replica becomes leader for term 10
- the leadership notification runs for term 10 and does a lookup of the
  translators, but one is still there from term 9, so the notification
  is a no-op
- stop_translator() task completes, removing the translator from term 9
- term 10 has no translator and never gets another notification to start
  a new one

This commit tweaks the code to make the notifications synchronously add
and remove translators from the map, rather than deferring that to
background work. The actual stopping of the translators is still
backgrounded, but that shouldn't be problematic.

andrwng commented:
```
/root/.cache/bazel/_bazel_root/a54ed064dc61dbb952fb48bf1f381b4a/sandbox/linux-sandbox/2623/execroot/_main/src/v/datalake/datalake_manager.cc:267:16: error: 't' used after it was moved [bugprone-use-after-move,hicpp-invalid-access-moved,-warnings-as-errors]
--
  | 267 \|         return t->stop().finally([_ = std::move(t)] {});
  | \|                ^
  | /root/.cache/bazel/_bazel_root/a54ed064dc61dbb952fb48bf1f381b4a/sandbox/linux-sandbox/2623/execroot/_main/src/v/datalake/datalake_manager.cc:267:39: note: move occurred here
  | 267 \|         return t->stop().finally([_ = std::move(t)] {});
  | \|                                       ^
  | /root/.cache/bazel/_bazel_root/a54ed064dc61dbb952fb48bf1f381b4a/sandbox/linux-sandbox/2623/execroot/_main/src/v/datalake/datalake_manager.cc:267:16: note: the use and move are unsequenced, i.e. there is no guarantee about the order in which they are evaluated
  | 267 \|         return t->stop().finally([_ = std::move(t)] {});
  | \|                ^
```

Same line use with a move isn't well defined. Pushed a fix

bharathv commented:
Is it better to explicitly ` // NOLINTNEXTLINE(bugprone-use-after-move)` ? otherwise it is unclear why there is a raw pointer access (for the reader)

andrwng commented:
> Is it better to explicitly ` // NOLINTNEXTLINE(bugprone-use-after-move)` ? otherwise it is unclear why there is a raw pointer access (for the reader)

Would prefer just adding a comment explaining the raw pointer. There are very few prod-code use-after-move lint escapes and we're probably better of that way

dotnwat commented on src/v/datalake/datalake_manager.cc:
```diff
@@ -260,14 +256,17 @@ void datalake_manager::start_translator(
     _translators.erase(it);
-    co_await translator->stop();
+    ssx::spawn_with_gate(_gate, [t = std::move(t)]() mutable {
+        auto* t_ptr = t.get();
+        return t_ptr->stop().finally([_ = std::move(t)] {});
```
what happens to the partition translator that is running if `_gate` is closed?

andrwng replied:
Any translator still in the map will be stopped in `datalake_manager::stop()`. Because we do the map update synchronously now in `stop_translator`, we should be guaranteed that the translator is stopped here or in the class-level stop

andrwng replied:
Oh actually I think that's not true still because `stop_translator` doesn't take the gate before updating the map

dotnwat replied:
> we should be guaranteed that the translator is stopped here or in the class-level stop

presumably the reference to it were just removed a few lines above

>    _translators.erase(it);

dotnwat replied:
yeh if you wrap more in the gate it would probably work ok

andrwng replied:
Yea, i'll add an early return (seems cleaner than just a gate hold + exception? at least, it's one fewer exception to worry about)

bharathv replied:
.. or early return in stop_translator() if gate is closed.

andrwng replied:
Done

Also added a note in the commit message of the new commit: i don't think this is problematic in practice, but still it seems easy enough to add anyway

andrwng replied:
>i don't think this is problematic in practice

I take it back, I bet there's a way for the `_iceberg_commit_interval.watch()` lambda to get unlucky and hit this since it spawns calls to `on_group_notification` in the background. It may may prevent the gate from being destructed with its hold, but it seems possible for its lambda to still be running after a call to `stop()` has been started. All the more reason to leave it in

andrwng committed:
datalake: opt out of stop_translator when stopping

Previously stop_translator() seemed to allow erased translators to go
un-stopped if the datalake_manager's gate is closed. This commit adds an
early return to avoid this.

It's not clear that this is actually problematic in practice because the
notification deregistration and gate close should happen atomically in
stop(), so it seems like notifications _should_ be guaranteed to run
without thinking about the gate. But it's an easy enough check to add
in.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59702-->
<!--c380d179-9b08-4176-a4dd-3ac8cdd6676f-->
<!--title-->
## Retry command for [Build#59702](https://buildkite.com/redpanda/redpanda/builds/59702)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/offset_for_leader_epoch_archival_test.py::OffsetForLeaderEpochArchivalTest.test_querying_archive
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--c9d9dacf-cddc-4dcc-bad8-5e734c722ad5-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59702">build#59702</a></summary>


| test_id                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_scrubber_test.CloudStorageScrubberTest.test_scrubber.cloud_storage_type=CloudStorageType.ABS | ducktape    | https://buildkite.com/redpanda/redpanda/builds/59702#0193bda8-0cfd-4c08-9e1f-8cfed822e6be | FLAKY         | 5/6      |
| rptest.tests.consumer_group_balancing_test.ConsumerGroupBalancingTest.test_coordinator_nodes_balance                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/59702#0193bda8-0d02-42bb-bd1e-2dbc3be6357b | FLAKY         | 5/6      |
| rptest.tests.offset_for_leader_epoch_archival_test.OffsetForLeaderEpochArchivalTest.test_querying_archive               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/59702#0193bda3-d009-498e-8636-86a16159144f | FAIL          | 0/1      |
</details>
<!--/body-->

vbotbuildovich commented:
/backport v24.3.x

rockwotj commented:
> The actual stopping of the translators is still backgrounded, but that shouldn't be problematic.

While I'm thinking about it - in Wasm the manager process all notifications in a single fiber (including shutdown of transform processors) to handle the case of a concurrent shutdown and startup of a processor. Is there any issues here with a translator concurrently being started and stopped? It's probably fine - I've not been in the details myself enough to know but figured I'd raise the question.

andrwng commented:
>Is there any issues here with a translator concurrently being started and stopped? It's probably fine - I've not been in the details myself enough to know but figured I'd raise the question.

Are you asking whether it's possible for a translator to be started while another translator is being stopped? With the current code, yes, but it shouldn't be problematic because the handoff of data to the coordinator accounts for concurrent translators/inconsistencies and rejects them.

Or if you're asking if it's possible for a single translator to have its start and stop race with each other, then the answer is no. `start_translator` is `void` -- it just schedules the background loop and adds the translator to the map, so there isn't really a window for a race to occur.

rockwotj commented:
> it shouldn't be problematic because the handoff of data to the coordinator accounts for concurrent translators/inconsistencies and rejects them.

This one - thanks for answering!
